framework/Optimizers/gradients/GradientApproximater.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2017 Battelle Energy Alliance, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+  TODO
+
+  Reworked 2020-01
+  @author: talbpaul
+"""
+import abc
+
+from utils import utils, InputData, InputTypes
+
+class GradientApproximater(utils.metaclass_insert(abc.ABCMeta, object)):
+  """
+    GradientApproximators use provided information to both select points
+    required to estimate gradients as well as calculate the estimates.
+  """
+  ##########################
+  # Initialization Methods #
+  ##########################
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Method to get a reference to a class that specifies the input data for class cls.
+      @ In, None
+      @ Out, specs, InputData.ParameterInput, class to use for specifying input of cls.
+    """
+    specs = InputData.parameterInputFactory(cls.__name__, ordered=False, strictMode=True)
+    specs.description = 'Base class for gradient approximation methods used in the GradientDescent Optimizer.'
+    return specs
+
+  @classmethod
+  def getSolutionExportVariableNames(cls):
+    """
+      Compiles a list of acceptable SolutionExport variable options.
+      @ In, None
+      @ Out, vars, dict, acceptable variable names and descriptions
+    """
+    return {}
+
+  def __init__(self):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    ## Instance Variable Initialization
+    # public
+    # _protected
+    self._optVars = None   # list(str) of opt variables
+    self._proximity = None # float, scaling for perturbation distance
+    self.N = None          # int, dimensionality
```

Comment:
self.N --> self._N?

Reply:
I don't know a reason why this change should be requested. `self.N` is used throughout the gradient approximators consistently, it looks like, and the choice to make the member public or protected seems largely subjective.

Reply:
I mentioned this, because this variable is listed under projected variable list.

Reply:
Oh, sure. I can move it.